runtime/src/main/java/org/corfudb/runtime/collections/DiskBackedCorfuTable.java

```diff
@@ -0,0 +1,505 @@
+                    "This is most likely an issue with the client's indexing function.", fatal);
+
+            close(); // Do not leave the table in an inconsistent state.
+
+            // In case of both a transactional and non-transactional operation, the client
+            // is going to receive UnrecoverableCorfuError along with the appropriate cause.
+            throw new UnrecoverableCorfuError(fatal);
+        }
+    }
+
+    public DiskBackedCorfuTable<K, V> clear() {
+        try {
+            rocksApi.clear();
+        } catch (RocksDBException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+        return this;
+    }
+
+    public Stream<Map.Entry<K, V>> entryStream() {
+        final RocksDbEntryIterator<K, V> entryIterator = rocksApi.getIterator(serializer);
+        Stream<Map.Entry<K, V>> resStream = StreamSupport.stream(
+                Spliterators.spliteratorUnknownSize(entryIterator, Spliterator.ORDERED), false);
+        return resStream.onClose(entryIterator::close);
+    }
+
+    public long size() {
+        if (persistenceOptions.getSizeComputationModel() == EXACT_SIZE) {
+            return rocksApi.exactSize();
+        }
+
+        return rocksApi.estimateSize();
+    }
+
+    public <I> Iterable<Map.Entry<K, V>> getByIndex(@NonNull final Index.Name indexName,
+                                                    @NonNull I indexKey) {
+        final List<ByteBuffer> keys = new ArrayList<>();
+        final List<ByteBuffer> values = new ArrayList<>();
+
+        try {
+            String secondaryIndex = indexName.get();
+            if (!secondaryIndexesAliasToPath.containsKey(secondaryIndex)) {
+                return null;
+            }
+
+            byte indexId = indexToId.get(secondaryIndexesAliasToPath.get(secondaryIndex));
+            rocksApi.prefixScan(columnFamilyRegistry.getSecondaryIndexColumnFamily(),
+                    indexId, indexKey, serializer, keys, values);
+
+            // Prevent the keys from being consumed.
+            final List<ByteBuffer> duplicateKeys = keys.stream().map(ByteBuffer::duplicate)
+                    .collect(Collectors.toList());
+            rocksApi.multiGet(columnFamilyRegistry.getDefaultColumnFamily(), duplicateKeys, values);
+
+            return Streams.zip(keys.stream(), values.stream(), (key, value) ->
+                    new AbstractMap.SimpleEntry<>(
+                            (K) serializer.deserialize(Unpooled.wrappedBuffer(key), null),
+                            (V) serializer.deserialize(Unpooled.wrappedBuffer(value), null)
+                    )).collect(Collectors.toList());
+        } catch (RocksDBException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+    }
+
+    @Override
+    public void close() {
+
+        // Do not call close on WriteOptions and Options, as they are
+        // either statically defined or owned by the client.
+
+        try {
+            this.rocksApi.close();
```

Comment:
This can cause a segfault.

One thread can delete the database instance while another thread is accessing it. For example, on a trimmed exception `RocksDB.destroyDB` can be called while another thread is consuming a scanning iterator.

Reply:
Fixed. When resetting the object, we now purge the MVO cache, which in turn will result in all snapshots associated with this object being released. I'll try to add a test as well.

Reply:
Test added.